collect_app/src/test/java/org/odk/collect/android/widgets/UrlWidgetTest.java

```diff
@@ -1,46 +1,123 @@
+
+    private CustomTabHelper customTabHelper;
+
+    @Before
+    public void setUp() {
+        customTabHelper = mock(CustomTabHelper.class);
     }
 
-    @NonNull
-    @Override
-    public StringData getNextAnswer() {
-        return new StringData(RandomString.make());
+    @Test
+    public void getAnswer_whenPromptAnswerDoesNotHaveAnswer_returnsNull() {
+        assertThat(createWidget(promptWithAnswer(null)).getAnswer(), nullValue());
     }
 
-    @Override
-    public void callingClearShouldRemoveTheExistingAnswer() {
-        // The widget is ReadOnly, clear shouldn't do anything.
+    @Test
+    public void getAnswer_whenPromptHasAnswer_returnsAnswer() {
+        UrlWidget widget = createWidget(promptWithAnswer(new StringData("blah")));
+        assertThat(widget.getAnswer().getDisplayText(), equalTo("blah"));
+    }
+
+    @Test
+    public void usingReadOnlyOption_makeAllClickableElementsDisabled() {
+        UrlWidget widget = createWidget(promptWithReadOnly());
+        Button urlButton = widget.findViewById(R.id.url_button);
+        assertThat(urlButton.getVisibility(), equalTo(View.GONE));
     }
 
     @Test
-    public void usingReadOnlyOptionShouldMakeAllClickableElementsDisabled() {
-        when(formEntryPrompt.isReadOnly()).thenReturn(true);
+    public void clearAnswer_doesNotClearWidgetAnswer() {
+        UrlWidget widget = createWidget(promptWithAnswer(new StringData("blah")));
+        widget.clearAnswer();
+        assertThat(widget.getAnswer().getDisplayText(), equalTo("blah"));
+    }
+
+    @Test
+    public void whenPromptHasAnswer_displaysAnswer() {
+        UrlWidget widget = createWidget(promptWithAnswer(new StringData("blah")));
+        assertThat(((TextView) widget.findViewById(R.id.url_answer_text)).getText().toString(), equalTo("blah"));
+    }
+
+    @Test
+    public void whenPromptAnswerDoesNotHaveAnswer_displayEmptyString() {
+        UrlWidget widget = createWidget(promptWithAnswer(null));
+        assertThat(((TextView) widget.findViewById(R.id.url_answer_text)).getText().toString(), equalTo(""));
+    }
+
+    @Test
+    public void clickingButtonWhenUrlIsEmpty_doesNotCallOpenUri() {
+        UrlWidget widget = createWidget(promptWithAnswer(null));
+        Button urlButton = widget.findViewById(R.id.url_button);
+        urlButton.performClick();
+
+        verify(customTabHelper, never()).bindCustomTabsService(null, null);
+        verify(customTabHelper, never()).openUri(null, null);
+    }
+
+    @Test
+    public void clickingButton_callsCorrectMethods() {
+        UrlWidget widget = createWidget(promptWithAnswer(new StringData("blah")));
+        Button urlButton = widget.findViewById(R.id.url_button);
+        urlButton.performClick();
+
+        verify(customTabHelper).bindCustomTabsService(widget.getContext(), null);
+        verify(customTabHelper).openUri(widget.getContext(), Uri.parse("blah"));
+    }
+
+    @Test
+    public void onLongPressButton_doesNotCallOpenUri() {
```

Comment:
I don't think this or the test below are correct. I think my question around writing tests for `setOnLongClickListener` and `cancelLongPress` was confusing though now I look at it. Lets breakdown what we need for each method:

* `setOnLongClickListener`: usually this needs to call `setOnLongClickListener` on any child views in the widget so that when I long press on anything in the widget I see the "Edit Prompt" pop up menu. Here you'll see that `setOnLongClickListener` is empty so that long pressing on the URL button won't do anything - it's clickable so it consumes the long press event. We need to write a test that makes sure long pressing on the url button will call any listener passed to the widget's `setOnLongClickListener`.
* `cancelLongPress`: this makes sure that any long press cancellation (usually the user scrolling around) works properly. I don't think there is a way to test that, so I think we should just make sure to review this method and make sure any views in the widget have the call propagated to them.

Reply:
One tip by the way for checking the coverage you have for an object is to use Android Studio's "Run with coverage" feature. If you use the green ▶️ icon to run the test file (in this case `URLWidgetTest`) and then hit "Run X with Coverage" you'll see green and red highligting in the `URLWidget` file for what lines are covered by tests and what ones aren't.

Reply:
Thank you so much for the suggestion. I will modify the tests.